aip/0194.md

```diff
@@ -0,0 +1,84 @@
+---
+aip:
+  id: 194
+  state: reviewing
+  created: 2019-08-16
+permalink: /194
+redirect_from:
+  - /0194
+---
+
+# Retries for gRPC clients
+
+RPCs sometimes fail. When one does, the client performing the RPC needs to know
+whether it is safe to retry the operation. When status codes are used
+consistently across multiple APIs, clients can respond to failures
+appropriately.
+
+## Guidance
+
+Clients **should** retry requests for which repeated runs would not cause
+unintended state changes, which are non-transactional, and which are either
+unary or server streaming.
+
+Clients **should not** retry transactional requests; instead these requests
+**should** have user-level retry logic that retries the entire transaction
+block from the start.
+
+Clients **should not** retry requests in which repeated runs would cause
+unintended state changes.
+
+**Note:** This AIP does not cover client streaming or bi-directional streaming.
+
+### Retryable codes
+
+The following codes **should** be retried for retryable requests:
+
+| Code | Discussion |
+|--|--|
+| `INTERNAL` | This generally means that some internal part of the system has failed. These are retryable under the expectation that the system will recover from failures. |
+| `UNAVAILABLE` | These codes generally result from network hiccups, and are generally transient. They are retryable under the expectation that the connection will become available (soon). |
+
+## Appendix
+
+### Non-retryable codes
+
+The following codes **should not** be retried for any request:
+
+| Code                | Discussion                                                                       |
+|---------------------|----------------------------------------------------------------------------------|
+| `OK`                | The request succeeded.                                                           |
+| `CANCELLED`         | A user can cancel a request, which **must** be honored.                          |
+| `DEADLINE_EXCEEDED` | A user can set a deadline, which **must** be honored.                            |
+| `INVALID_ARGUMENT`  | Retrying a request with an invalid argument will never succeed.                  |
+| `DATA_LOSS`         | This is an unrecoverable error and **must** immediately be surfaced to the user. |
+
+### Generally non-retryable codes:
+
+The following codes generally **should not** be retried for any request:
+
+| Code  | Discussion  |
+|--|--|
+| `RESOURCE_EXHAUSTED` | This may be a signal that the user's quota is exhausted. Retries therefore may not be expected to work for several hours; meanwhile the retries may have billing implications. If `RESOURCE_EXHAUSTED` is used for other reasons than quota and the expected time for the resource to become available is much shorter, it may be retryable. |
```

Comment:
We may wish to add something around a machine-readable status that indicates the nature of the exhausted resource, and when it will become available again. That would definitely be nice to have - but unless we've got something concrete, it possibly makes sense to leave it out. (I'd sort of like "this is reasonable, but not specified yet - we're aware of it" somehow.)

Reply:
+1. I unfortunately don't have any concrete for that, though.